Scenario Editor (Part 1)

Noordfrees opened the pull request:
Extends the editor to allow placing and in-detail-configuring buildings, flags, roads, tribe immovables, ships, workers etc, and setting player relations, allowed building types, territories, and vision.

Generalizes a lot of game-specific functionality for use in the editor.

Provides the baseplate for writing a mission_thread without any knowledge of Lua. The GUI for this will come in a follow-up branch.

Fixes #651 
Fixes #2989 
closes #1782

Commit:
Started implementing scenario-creating features into the editor

Commit:
Create nicer init.lua

Commit:
Added infrastructure tools

Commit:
Editor saveloading "fix"

Commit:
Started to implement building settings in editor

Commit:
Fixed all the compile errors

Commit:
Undid accidental changes

Commit:
Removed unneeded dynamic_casts

Commit:
Infrastructure tool window cleanup

Noordfrees committed:
Intermediate commit

Noordfrees committed:
Merged master

Noordfrees committed:
Implemented settings for inputqueues and productionsite workers. Fixed compile errors.

Noordfrees committed:
Merged master

Noordfrees committed:
Merged master

Noordfrees committed:
Implemented soldier settings and fixed an assert in productionsitewindow

Noordfrees committed:
Intermediate commit

Noordfrees committed:
Merged master

Noordfrees committed:
Fixed merging errors

Noordfrees committed:
Implemented worker tool window. Some fixes.

Noordfrees committed:
Implemented road tool

Noordfrees committed:
Fixed road tool CTRL handling and info message

Noordfrees committed:
Merged master

Noordfrees committed:
Fixed place-infra-tool handling of force and csite

Noordfrees committed:
Some fixes for place-road, delete-road, place-infra, delete-infra, and infra-settings tools

Noordfrees committed:
Worker deletion fixes and warehousewindow-adjustment

Noordfrees committed:
Implemented scenario difficulty stages

Noordfrees committed:
Started to implement scripting support

Noordfrees committed:
Merged master

Noordfrees committed:
Made scripting objects classes. Cleanup lua includes and finalizing detection.

Noordfrees committed:
Continued implementing scripting support

Noordfrees committed:
Minor cleanup

Noordfrees committed:
Added all builtin lua functions

Noordfrees committed:
Fix game-startup issues

Noordfrees committed:
Implemented player team and relation settings

Noordfrees committed:
Merged master

Noordfrees committed:
Split scripting stuff into several files and added builtin properties

Noordfrees committed:
Formatting

Noordfrees committed:
Added functions and properties of module interfaces

Noordfrees committed:
Fixed make infra tool building placement

Noordfrees committed:
Implemented logical and mathematical operators

Noordfrees committed:
Simplified ScriptingObject loading and deleting of unused objects

Noordfrees committed:
Implemented while loops and if clauses

Noordfrees committed:
Fixed start-of-game buildings and road init

Noordfrees committed:
Implemented table access, operators '==nil' and '..', and for-each loops

Noordfrees committed:
Implemented for-each and fixed VariableType inheriting

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into scenario-editor

Noordfrees committed:
Added modulo operator

Noordfrees committed:
Implemented allowed buildings tool. Save teams etc in the map.

Noordfrees committed:
Implemented vision tool

Noordfrees committed:
Implemented scriptingobject selftest and return/break statements

Noordfrees committed:
Split editor into multiple libs

Noordfrees committed:
Improved workers tool. New ship configuration window.

Noordfrees committed:
Refactor worker tool window and shipping window

Noordfrees committed:
Refactored ControlStructures and implemented ConstructionSiteWindow functionality

Noordfrees committed:
Fixed Lua indentation handling and improved some windows

Noordfrees committed:
Refactored infra placement tool

Noordfrees committed:
Properties and VariableTypeID cleanup

Noordfrees committed:
Cleanup

Noordfrees committed:
Infra window layouting fix

Noordfrees committed:
Merge branch 'master' into scenario-editor

Noordfrees committed:
Merge branch 'master' into scenario-editor

Noordfrees committed:
ran clang-format

Noordfrees committed:
clang-format does nothing, so let travis tell me what's wrong

Noordfrees committed:
Manually corrected code formatting

Noordfrees committed:
Fixed editor test case compatibility

Noordfrees committed:
Merge branch 'master' into scenario-editor

Noordfrees committed:
Making travis happy

Noordfrees committed:
Making travis happy

Noordfrees committed:
Merge branch 'master' into scenario-editor

gunchleoc commented:
I gave it a quick spin, excellent feature :)

Some problems to solve:

1. I think this should produce Lua code rather than binary map packets. Otherwise, we will have to provide savegame compatibility forever down to this version, because scenarios will break. At the moment, we can clean up the ugly compatibility spaghetti code after each release to contain map compatibility only. This advantage will be gone.
2. I had some trouble placing flags, and I could not place roads at all. This could be done in a follow-up branch though.
3. Wishlist: add buttons to center on each player's headquarters to make jumping between players easier.

Noordfrees commented:
1) The editor does write lua code on every save. (`EditorInteractive::write_lua()`). There just isn't any mission thread written yet. (Should I add some placeholder like `print("Hello World")`?). MapScenarioEditorPacket is there only for the editor. In games, it only does some post-loading initialization which is not read from file and therefore does not even consider the packet version.
2) Could you describe what exactly you did so I know whether there's a bug or whether the correct usage is just not as intuitive as I believe?
3) Will implement

Noordfrees committed:
Merge branch 'master' into scenario-editor

gunchleoc commented:
1. I'm not talking about the mission_thread, I'm talking about buildings, roads and workers like in https://github.com/widelands/widelands/blob/master/data/campaigns/emp04.wmf/scripting/starting_conditions.lua
2. I just tried to place a few flags. I did deactivate the force option and then it became unintuitive. I think it should behave just like in-game, otherwise it will be hard to learn. I did not figure out how to place roads at all, maybe yo didn't write the code for that yet.

Noordfrees commented:
1. Yes, this is not done via Lua because placing buildings in the script is much slower. It easily takes several minutes for a larger settlement. But this is saved in the map exactly in the same way it would be in a real game.
2. When placing flags without `force` it should work if and only if the place is owned by the player and there are no obstacles or nearby flags.
Building a road is done like this:
- Click on the flag from which you want to start
- Now you're in roadbuilding mode and the slope overlays etc are visible
- Build the road just like in the game
- Click on a flag to build the road up to there, or click the end field again to build a flag there and complete the road if allowed.

gunchleoc commented:
1. We need to write a more efficient Lua script then. I really really don't want to have to deal with savegame compatibility for this because that code is just ugh and hell to debug already.

2. I still can't enter road building mode. Automatic road building mode is on in my options, so that's not the problem

Noordfrees commented:
1. You mean, we ignore buildings/flags/roads/workers etc during editor saveloading, and save all this info in MapScenarioEditorPacket instead (for reloading the scenario in the editor) and additionally write it into the Lua script? In that case I'd like to give the user an option "Save infrastructure in the map. This is faster during startup but may cause your scenario to be incompatible with future Widelands releases."
2. Automatic roadbuild mode is not considered by the editor. Nor is this possible, because flags and roads are placed with different tools:
Place a flag with the infrastructure tool, then select the road tool, then click on a flag and build the road.

gunchleoc commented:
1. No, don't save it to the map. Lua code only, and no option - if we have an option, we're back at the problem that we have to provide savegame compatibility ad nauseam. We have code to create Lua Table source code, you could take that as a starting point for prodiving Lua tables for bulk building placement functions. https://github.com/widelands/widelands/tree/master/src/website/lua
2. OK, I didn't find the road tool then. How about extending the in-game field actions with the additional editor objects? That might be easier to learn for the user.

And I found another issue - when clicking on "Play Tutorial" in the main menu, I get

```
==15933==ERROR: AddressSanitizer: SEGV on unknown address 0x0000000000e8 (pc 0x5561ff9ec136 bp 0x7fffc28d3830 sp 0x7fffc28d3760 T0)
==15933==The signal is caused by a READ memory access.
==15933==Hint: address points to the zero page.
    #0 0x5561ff9ec135 in std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::end() const /usr/include/c++/7/bits/stl_vector.h:591
    #1 0x5561ff9ec6ee in std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::empty() const /usr/include/c++/7/bits/stl_vector.h:760
    #2 0x5561ffda6339 in FullscreenMenuScenarioSelect::FullscreenMenuScenarioSelect(CampaignData*) ../src/ui_fsmenu/scenario_select.cc:116
    #3 0x5561ff7e82f5 in WLApplication::mainmenu_tutorial() ../src/wlapplication.cc:1193
    #4 0x5561ff7e7a2d in WLApplication::mainmenu() ../src/wlapplication.cc:1128
    #5 0x5561ff7d8257 in WLApplication::run() ../src/wlapplication.cc:493
    #6 0x5561ff7d3a3e in main ../src/main.cc:44
    #7 0x7fa4e4af9b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
    #8 0x5561ff7d38b9 in _start (/media/cuideigin/Linux/sources/widelands/github/widelands/widelands+0x109d8b9)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /usr/include/c++/7/bits/stl_vector.h:591 in std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::end() const
==15933==ABORTING

```

Noordfrees committed:
Merged master

Noordfrees committed:
Fixed tutorial ui crash

Noordfrees committed:
Merged master

actions-user committed:
9 files were automatically formatted.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into scenario-editor

gunchleoc commented:
I noticed that placing buildings & flags/roads via Lua generates a lot of log output. If we omitted that, we should get a speed up.

Noordfrees committed:
Started to address code review...

Noordfrees committed:
Merge branch 'master' into scenario-editor

actions-user committed:
4 files were automatically formatted.

Noordfrees committed:
Merge branch 'master' into scenario-editor

actions-user committed:
2 files were automatically formatted.

Noordfrees committed:
Continued addressing review

Noordfrees committed:
Merge branch 'scenario-editor' of github.com:Noordfrees/widelands into scenario-editor

Noordfrees committed:
Refactored road/ww building

Noordfrees committed:
Added go to player startpos buttons to player teams menu

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into scenario-editor

actions-user committed:
12 files were automatically formatted.

Noordfrees committed:
making travis happy

actions-user committed:
'src/logic/map.cc' was automatically formatted.

Noordfrees committed:
Fix testsuite

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into scenario-editor

Noordfrees committed:
Fix testsuite

Noordfrees committed:
Continued addressing review

actions-user committed:
2 files were automatically formatted.

Noordfrees committed:
Continued addressing review

actions-user committed:
4 files were automatically formatted.

Noordfrees committed:
Continued addressing review

actions-user committed:
2 files were automatically formatted.

Noordfrees committed:
Merge branch 'master' into scenario-editor

hessenfarmer commented:
just gave it a try.
1. when setting view the map should be fully visible again after closing the tool.   
2. when trying to save lua files on windows it crashes without any useful output.   
3. placing of flags should be in roads dialog rather then in infrastructure
4. setting ownership is difficult as only every second point shown is accesible. please show only draggable points.
5. suggestion have a concentric Ownership tool (by pressing CTRL f.e. pressing CTRL and moving a point should make a circular ownership with the former point as center.  
6. Suggestion make an overlay for the the owned territory
7. make it possible to set an HQ on starting field
8. suggestion when placing a road make the same option to autoplace flags like in the normal game
9. I couldn't acces any dialog for the settings (shift+I)

Noordfrees commented:
1) will look into this
2) is to be expected ATM, I'm in the middle of addressing the code review and still have some nasty new saveloading bugs :(
3) Would be pretty difficult because in my current implementation the roads tool is closely linked to entering roadbuilding mode. Having two modes road/flag here would introduce all kind of new difficulties…
4) 
5) I don't understand these two points could you explain more? In the Ownership options menu, you choose the player for whom you want to claim nodes with the dropdown, and then you can set the tool size with the keys 1-0 and click to change the owner of fields
6) will do
7) ok, will add a convenience button to the place infrastructure tool to place appropriate HQs for all players at their starting points
8) Implemented in theory but unfortunately the editor consumes shift and ctrl modifiers to switch to alternative tools. So I don't see how this could be done cleanly – perhaps as another checkbox in the tool window?
9) Press Shift+I to activate the tool, then click on a building, flag, or ship to edit

Noordfrees committed:
Addressed review

Noordfrees committed:
Merged master

actions-user committed:
4 files were automatically formatted.

Noordfrees committed:
Add field ownership overlay

actions-user committed:
3 files were automatically formatted.

Noordfrees committed:
Added missing builtins

Noordfrees committed:
Refactored builtin saving

actions-user committed:
'src/editor/scripting/builtin.cc' was automatically formatted.

Noordfrees committed:
Fixed init.lua generation

actions-user committed:
'src/editor/editorinteractive.cc' was automatically formatted.

Noordfrees committed:
Fixed scenario packet saveloading

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into scenario-editor

Noordfrees commented:
I think I addressed all the review comments now

Noordfrees committed:
Merged master

Noordfrees committed:
boost::function → std::function

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into scenario-editor

Noordfrees committed:
Merge branch 'master' into scenario-editor

Noordfrees committed:
Merge branch 'master' into scenario-editor

Noordfrees committed:
Merged master

Noordfrees committed:
Merge branch 'scenario-editor' of github.com:Noordfrees/widelands into scenario-editor

Noordfrees committed:
Compile eror

Noordfrees committed:
Remove superfluous includes and forward declarations

actions-user committed:
'src/editor/ui_menus/player_teams_menu.h' was automatically formatted.

Noordfrees committed:
Merge branch 'master' into scenario-editor

hessenfarmer commented:
ok I tested this again. tools work good and after the explanations regarding the radius of a tool it was possible to establish a swift workflow to place buildings make roads and edit settings of the buildings. 
However I encountered following points:
- suggestion is to have a GUI element for the tool radius (together with a tooltip for the numbers)    
- while setting workers the experience buttons are clickable even if no worker assigned. if you click them the editor crashes. They should only be clickable if a worker can gain experience. 
- I could not save a map it crashes while trying to write a lua file. This might be a windows problem perhaps
- if we merge this we should disable scripting as it is not yet implemented.

Noordfrees committed:
Merged master

Noordfrees committed:
Addressed review

Noordfrees commented:
> suggestion is to have a GUI element for the tool radius (together with a tooltip for the numbers)

This is exactly the same as for all other editor tools: Use the same main-toolbar button you use for all other tools as well.

> while setting workers the experience buttons are clickable even if no worker assigned. if you click them the editor crashes. They should only be clickable if a worker can gain experience.

Cannot reproduce, for me the button is enabled only when a worker type that requires experience is selected…

> I could not save a map it crashes while trying to write a lua file. This might be a windows problem perhaps

Cannot reproduce this either, it creates a nice init.lua with a few thousand lines of initialization code. Perhaps this might have something to do with `\n`/`\r\n` (or whatever windows uses) confusion? Do you have an stdout or a backtrace?

actions-user committed:
'src/editor/editorinteractive.cc' was automatically formatted.

hessenfarmer commented:
> 
> 
> > suggestion is to have a GUI element for the tool radius (together with a tooltip for the numbers)
> 
> This is exactly the same as for all other editor tools: Use the same main-toolbar button you use for all other tools as well.   
  
what I meant was for example in the height tool you could set the height via keyboard or with a GUI Element such a thing would be good for the vision and the ownership tool as well.   

> 
> > while setting workers the experience buttons are clickable even if no worker assigned. if you click them the editor crashes. They should only be clickable if a worker can gain experience.
> 
> Cannot reproduce, for me the button is enabled only when a worker type that requires experience is selected…    
   
if you build for example a quarry and then use settings there is no worker selected in the beginning. However the experience settings are active. if you click them it crashes.    
   
> 
> > I could not save a map it crashes while trying to write a lua file. This might be a windows problem perhaps
> 
> Cannot reproduce this either, it creates a nice init.lua with a few thousand lines of initialization code. Perhaps this might have something to do with `\n`/`\r\n` (or whatever windows uses) confusion? Do you have an stdout or a backtrace?   
  
Stdout ends with:
writing LUA files...
could you point me to the code to write the lua file?

Noordfrees commented:
> what I meant was for example in the height tool you could set the height via keyboard or with a GUI Element such a thing would be good for the vision and the ownership tool as well.

I think we are talking at cross-purpose… You can use the hotkeys 1-0 for the size of any tool, including height, resources, terrain, ownership, vision, etc. And there is a button "Tool Size" in the bottom toolbar which brings up a small window with two buttons to change the toolsize, and this is also available for any tool, be it height, terrain, ownership, vision or whatever.  
The new tools behave exactly like the existing tools here, except that they are located in a different dropdown.

> if you build for example a quarry and then use settings there is no worker selected in the beginning. However the experience settings are active. if you click them it crashes.

Ah I thought you meant the Workers&Ships tool. Will investigate…

>  could you point me to the code to write the lua file?

Beware of huge function ahead ;) https://github.com/Noordfrees/widelands/blob/scenario-editor/src/editor/editorinteractive.cc#L1441

Noordfrees committed:
Initially disable experience buttons in productionsitewindow

hessenfarmer commented:
did some further testing. could save a map with scenario as long as no road was built. so the culprit might be in the road section.

hessenfarmer commented:
> > what I meant was for example in the height tool you could set the height via keyboard or with a GUI Element such a thing would be good for the vision and the ownership tool as well.
> 
> I think we are talking at cross-purpose… You can use the hotkeys 1-0 for the size of any tool, including height, resources, terrain, ownership, vision, etc. And there is a button "Tool Size" in the bottom toolbar which brings up a small window with two buttons to change the toolsize, and this is also available for any tool, be it height, terrain, ownership, vision or whatever.
> The new tools behave exactly like the existing tools here, except that they are located in a different dropdown.    
   
ok didn't realize the little thing in the bottom line everything ok then except we should probably initialize with a toolsize of 2 or even 3 for the ownership tool. 
>

Noordfrees committed:
fix crash when road has less carriers than possible

Noordfrees commented:
> did some further testing. could save a map with scenario as long as no road was built. so the culprit might be in the road section.

Found and fixed it :)

> ok didN't realize the little thin in the bottom line everything ok then except we should probably initialize with a toolsize of 2 or even 3 for the ownership tool.

It is initialized with the size of the previously selected tool – consistently like every other tool in the editor…

hessenfarmer commented:
ok now it seems everything is working. only thing left is that the place headquarters button is always placing all HQ this should be per player I believe

Noordfrees committed:
Tweak auto-infra button

hessenfarmer reviewed (changes requested):
i was trying to play the current scenarios. after starting emp04 and clicking on a lumberjacks building it crashed so I assume this breaks compatibility with existing scenarios where infrastructure has been placed. From test currently emp04 and the seafaring warfare and economy tutorial are at least affected.

Noordfrees commented:
It cannot break scenario compatibility. The crash is most likely related to the changes to the building windows because I'm reusing the existing UI for the editor. Will investigate…

Noordfrees committed:
Fixed buildingwindow crash

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into scenario-editor

Noordfrees commented:
Simple oversight, should be fixed

hessenfarmer reviewed (approved):
ok from my side this is good to go. However I couldn't review the code as I don't understand enough c++ to do so. 
But from testing this should be green now.
thanks for this extraordinary feature and the new possibilities in scripting as well.

Noordfrees commented:
Thanks for the review :) I'd like to have a code review as well though as this change is pretty huge and also touches a lot of existing game functionality.

gunchleoc reviewed (changes requested):
Here's my testing result:

Undo removing a building results in

```
==11837==The signal is caused by a READ memory access.
==11837==Hint: address points to the zero page.
    #0 0x55df163008cb in Widelands::MapObject::remove(Widelands::EditorGameBase&) ../src/logic/map_objects/map_object.cc:468
    #1 0x55df1621e555 in ScenarioInfrastructureTool::handle_undo_impl(Widelands::NodeAndTriangle<Widelands::Coords, Widelands::Coords> const&, EditorInteractive&, EditorActionArgs*, Widelands::Map*) ../src/editor/tools/scenario_infrastructure_tool.cc:87
    #2 0x55df1585597e in EditorTool::handle_undo(EditorTool::ToolIndex, Widelands::NodeAndTriangle<Widelands::Coords, Widelands::Coords> const&, EditorInteractive&, EditorActionArgs*, Widelands::Map*) ../src/editor/tools/tool.h:61
    #3 0x55df15853e3c in EditorHistory::undo_action() ../src/editor/tools/history.cc:76
    #4 0x55df157af4fb in EditorInteractive::handle_key(bool, SDL_Keysym) ../src/editor/editorinteractive.cc:1181
    #5 0x55df15c428d4 in UI::Panel::do_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:923
    #6 0x55df15c4325c in UI::Panel::ui_key(bool, SDL_Keysym) ../src/ui_basic/panel.cc:1073
    #7 0x55df157468f3 in WLApplication::handle_input(InputCallback const*) ../src/wlapplication.cc:620
    #8 0x55df15c3cadc in UI::Panel::do_run() ../src/ui_basic/panel.cc:182
    #9 0x55df1576ab47 in UI::Panel::Returncodes UI::Panel::run<UI::Panel::Returncodes>() ../src/ui_basic/panel.h:101
    #10 0x55df157b0763 in EditorInteractive::run_editor(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ../src/editor/editorinteractive.cc:1286
    #11 0x55df1575446d in WLApplication::mainmenu() ../src/wlapplication.cc:1140
    #12 0x55df15744b4b in WLApplication::run() ../src/wlapplication.cc:483
    #13 0x55df1574033e in main ../src/main.cc:44
    #14 0x7f7f84aedb96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
    #15 0x55df157401b9 in _start (widelands+0x107f1b9)
```

1. The ware type a worker can carry should be restricted to their buildings' input/output except for carriers, but that can be done in a follow-up branch.
2. Ship has no menu icon in the "Workers and Ships" tool
3. The attached map produces `lua: [../src/scripting/lua_errors.cc:22] [string "scripting/init.lua"]:22: bad argument #1 to 'place_road' (string expected, got nil)` 
[Scenario editor.wmf.zip](https://github.com/widelands/widelands/files/4322830/Scenario.editor.wmf.zip)

Noordfrees committed:
Addressed review

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into scenario-editor

Noordfrees commented:
2. and 3. are fixed now. 
Fixed the crash by disallowing undo for most scenario tools (undoing delete operations is not possible because it can have a number of ugly side effects; the crash was caused by undoing the placing of an already deleted object).

gunchleoc commented:
Looks good to me. I'll try to do a code review soonish.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into scenario-editor

Noordfrees committed:
Merged master

actions-user committed:
4 files were automatically formatted.

Noordfrees committed:
Added missing new builtins

gunchleoc commented:
I have started a code review, but I'll have to finish another day.

Note: All the scenarios need retesting

For the "Choose a Scenario" screen, I'd add "Difficulty" as a header and remove the label from the dropdown. We need to keep some room in the dropdown for translations.

Noordfrees committed:
Addressed code review

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into scenario-editor

Noordfrees committed:
Adressed review

actions-user committed:
3 files were automatically formatted.

Noordfrees committed:
Moved scenario selection difficulty string to a textarea

actions-user committed:
'src/ui_fsmenu/scenario_select.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'master' into scenario-editor

Noordfrees committed:
A round of addressing code review …

actions-user committed:
3 files were automatically formatted.

Noordfrees committed:
Continued addressing review

Noordfrees committed:
Merge branch 'scenario-editor' of github.com:Noordfrees/widelands into scenario-editor

Noordfrees committed:
Continued addressing review …

actions-user committed:
'src/wui/constructionsitewindow.cc' was automatically formatted.

Noordfrees committed:
Continued addressing review …

actions-user committed:
'src/scripting/lua_map.cc' was automatically formatted.

Noordfrees committed:
Continued addressing review …

gunchleoc commented:
I'm done. Looking forward to this feature :)

Testing still needed:

* We need to retest the scenarios because of the changes to road building in Lua
* The road and waterway building code has been refactored, so that needs testing too
* Building windows in game and editor modes

Some general notes:

* Please address my comments in `editor/scripting`, then rip it out and stick it in a follow-up branch. If I understand correctly, this code will be needed for `mission_thread` but is unused right now. This is simply too much code to review and merge while it's still completely untested. So, I gave up on going through all of it.
* The scenario difficulty is an unrelated feature, please create a separate pull request for that part of the code so that we will have a clean commit history. No need to rip this out, we can simply merge the other branch first. I haven't done a code review on this yet.
* Refactoring of waterways/road building also looks like a separate feature that should have its own pull request. We can also merge that one before this branch so you won't have to change the code here.

And a bug while clicking with the info tool on a worker:

```
Error rendering richtext: Syntax error at 1:2370: expected an allowed attribute for '' tag, got 'author<vspace gap'. String continues with: '=6></font></p></div></div><div width=100%><div><p><font size=14 color=ffff00>•</font></p></div><di'. Text is:
<rt><p><font face=sans size=14 color=d1d1d1 bold=1>Node</font></p><div width=100%><div><p><font size=14 color=ffff00>•</font></p></div><div><p><space gap=6></p></div><div width=*><p><font size=14 color=ffff00>Coordinates: (6, 10)<vspace gap=6></font></p></div></div><div width=100%><div><p><font size=14 color=ffff00>•</font></p></div><div><p><space gap=6></p></div><div width=*><p><font size=14 color=ffff00>Caps: big, flag, walkable<vspace gap=6></font></p></div></div><div width=100%><div><p><font size=14 color=ffff00>•</font></p></div><div><p><space gap=6></p></div><div width=*><p><font size=14 color=ffff00>Owned by: —<vspace gap=6></font></p></div></div><p><vspace gap=6><font face=sans size=14 color=d1d1d1 bold=1>Terrain</font></p><div width=100%><div><p><font size=14 color=ffff00>•</font></p></div><div><p><space gap=6></p></div><div width=*><p><font size=14 color=ffff00>Name: Meadow 1<vspace gap=6></font></p></div></div><div width=100%><div><p><font size=14 color=ffff00>•</font></p></div><div><p><space gap=6></p></div><div width=*><p><font size=14 color=ffff00>Is: arable<vspace gap=6></font></p></div></div><div width=100%><div><p><font size=14 color=ffff00>•</font></p></div><div><p><space gap=6></p></div><div width=*><p><font size=14 color=ffff00>Editor Category: Summer<vspace gap=6></font></p></div></div><p><vspace gap=6><font face=sans size=14 color=d1d1d1 bold=1>Objects</font></p><div width=100%><div><p><font size=14 color=ffff00>•</font></p></div><div><p><space gap=6></p></div><div width=*><p><font size=14 color=ffff00>Workers: Brewer (Barbarians), XP: 0/10<vspace gap=6></font></p></div></div><p><vspace gap=6><font face=sans size=14 color=d1d1d1 bold=1>Map</font></p><div width=100%><div><p><font size=14 color=ffff00>•</font></p></div><div><p
```

Noordfrees committed:
Continued addressing review …

Noordfrees committed:
Continued addressing review …

Noordfrees commented:
> Some general notes:
> 
>     * Please address my comments in `editor/scripting`, then rip it out and stick it in a follow-up branch. If I understand correctly, this code will be needed for `mission_thread` but is unused right now. This is simply too much code to review and merge while it's still completely untested. So, I gave up on going through all of it.

This code is already being used. The EIA has some variables of type `LuaFunction`, `FS_LocalVarDeclOrAssign` and `FS_LaunchCoroutine` which require all the new functionality. I had previously hardcoded some tests to check that this works fine and later removed them. Now the `LuaFunction` "mission_thread" contains some scripting function (editorinteractive.cc line 1366-1382). So this functionality is needed for this branch already.

>     * The scenario difficulty is an unrelated feature, please create a separate pull request for that part of the code so that we will have a clean commit history. No need to rip this out, we can simply merge the other branch first. I haven't done a code review on this yet.
> 
>     * Refactoring of waterways/road building also looks like a separate feature that should have its own pull request. We can also merge that one before this branch so you won't have to change the code here.

Ok, will do so…

> And a bug while clicking with the info tool on a worker:

Can't reproduce. Which worker and which map?

actions-user committed:
2 files were automatically formatted.

Noordfrees committed:
Continued addressing review …

actions-user committed:
4 files were automatically formatted.

Noordfrees committed:
Continued addressing review

Noordfrees committed:
Continued addressing review

actions-user committed:
'src/scripting/lua_map.cc' was automatically formatted.

gunchleoc commented:
> > Some general notes:
> > ```
> > * Please address my comments in `editor/scripting`, then rip it out and stick it in a follow-up branch. If I understand correctly, this code will be needed for `mission_thread` but is unused right now. This is simply too much code to review and merge while it's still completely untested. So, I gave up on going through all of it.
> > 
> 
> This code is already being used. The EIA has some variables of type `LuaFunction`, `FS_LocalVarDeclOrAssign` and `FS_LaunchCoroutine` which require all the new functionality. I had previously hardcoded some tests to check that this works fine and later removed them. Now the `LuaFunction` "mission_thread" contains some scripting function (editorinteractive.cc line 1366-1382). So this functionality is needed for this branch already.

OK, I'll need to find some time to review this too then. Hopefully next weekend.

> > And a bug while clicking with the info tool on a worker:
> 
> Can't reproduce. Which worker and which map?

New map. barbarian_brewer if I remember correctly. I just placed a few workers for testing the experience/carry tool functions.

Noordfrees commented:
Still can't reproduce. What I could imagine is that it is somehow related to your map's author or description string (or i18n-related). E.g. the info tool window is messed up for LongLongWay ('>' and '<' in map description) which seems to be a similar problem as #3760

Noordfrees committed:
Refactored builtin.cc to use std::map for faster lookup

Noordfrees committed:
Merge branch 'scenario-editor' of github.com:Noordfrees/widelands into scenario-editor

actions-user committed:
2 files were automatically formatted.

Noordfrees committed:
Created testcase and fixed a typo

Noordfrees committed:
Continued addressing review

actions-user committed:
'src/wui/soldierlist.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'master' into scenario-editor

Noordfrees committed:
Add lua handling for invalid warehouse policies

gunchleoc commented:
> Still can't reproduce. What I could imagine is that it is somehow related to your map's author or description string (or i18n-related). E.g. the info tool window is messed up for LongLongWay ('>' and '<' in map description) which seems to be a similar problem as #3760

You're right, the bug is in master. I have created a PR with a fix: https://github.com/widelands/widelands/pull/3798

Noordfrees committed:
Completed addressing review.

actions-user committed:
'src/wui/inputqueuedisplay.cc' was automatically formatted.

Noordfrees committed:
Open building windows after auto-infra

Noordfrees committed:
Merge branch 'scenario-editor' of github.com:Noordfrees/widelands into scenario-editor

actions-user committed:
'src/editor/ui_menus/scenario_tool_infrastructure_options_menu.cc' was automatically formatted.

Noordfrees committed:
fix testsuite

Noordfrees committed:
Merged master

actions-user committed:
'src/wui/interactive_base.h' was automatically formatted.

Noordfrees committed:
Commented out unused code and moved some code around

Noordfrees committed:
Merged master

Noordfrees committed:
Merge branch 'master' into scenario-editor

Noordfrees committed:
Fixed a regressed crash

Noordfrees committed:
Merge branch 'scenario-editor' of github.com:Noordfrees/widelands into scenario-editor

Noordfrees committed:
Merge branch 'master' into scenario-editor

Noordfrees committed:
Merged master

Noordfrees committed:
Merge branch 'scenario-editor' of github.com:Noordfrees/widelands into scenario-editor

Noordfrees commented:
Rescheduling for b22 since I can think of a couple of potential complications which we do not want to have in b21.

Noordfrees committed:
Merged master

actions-user committed:
2 files were automatically formatted.

Noordfrees committed:
Merged master

actions-user committed:
10 files were automatically formatted.

Noordfrees committed:
Making travis happy

Noordfrees committed:
Merged master

actions-user committed:
13 files were automatically formatted.

Noordfrees committed:
Merged master

actions-user committed:
'src/economy/economy.cc' was automatically formatted.

Noordfrees committed:
typo

actions-user committed:
'src/economy/economy.cc' was automatically formatted.

Noordfrees committed:
Merged master and fixed RST formatting

Noordfrees committed:
Merge branch 'master' into scenario-editor

Noordfrees committed:
Started to merge master…

actions-user committed:
5 files were automatically formatted.

gunchleoc reviewed (commented):
I have added a few nits. This branch is too big to be reviewed properly, so we'll have to just merge it in and deal with the fallout later.

Please make sure that it compiles (which it currently doesn't) and that it won't bug out while playing a game / editing maps without the new bells & whistles.

After this branch is in, we should also compare with https://github.com/widelands/widelands/tree/master/src/website/lua so that we will end up with 1 implementation for writing Lua tables to code instead of 2.

Noordfrees committed:
First round of compile errors

Noordfrees committed:
Merged remote branch

Noordfrees committed:
Merged master

Noordfrees committed:
Fixed compile errors

actions-user committed:
6 files were automatically formatted.

Noordfrees committed:
Make clang-tidy happy

Noordfrees committed:
Merged master

Noordfrees committed:
Delete an unused variable

Noordfrees committed:
Clang-tidy found another nit

gunchleoc commented:
I noticed that the test suite failed on an empty object name. Maybe calling `EditorGameBase::load_all_tribes()` somewhere can help you there.

Noordfrees committed:
Merged master

Noordfrees committed:
Fixing the editor testsuite

actions-user committed:
'src/logic/playersmanager.cc' was automatically formatted.

Noordfrees committed:
Merged master

gunchleoc commented:
@Noordfrees When you think this is ready again, please post here and I'll do a final round of testing.

Noordfrees committed:
Merged master

actions-user committed:
7 files were automatically formatted.

Noordfrees commented:
Saveloading scenario-specific data is currently broken. Hoping to get this ready soonish…

Noordfrees commented:
This branch has become unmaintainable. I'll make some smaller refactoring branches soonish so at least parts of the changes here can be properly checked again.
